src/signet.cpp

```diff
@@ -0,0 +1,147 @@
+            }
+            replacement << pushdata;
+        } else {
+            replacement << opcode;
+        }
+    }
+
+    if (found) script = replacement;
+    return found;
+}
+
+static uint256 ComputeModifiedMerkleRoot(const CMutableTransaction& cb, const CBlock& block)
+{
+    std::vector<uint256> leaves;
+    leaves.resize(block.vtx.size());
+    leaves[0] = cb.GetHash();
+    for (size_t s = 1; s < block.vtx.size(); ++s) {
+        leaves[s] = block.vtx[s]->GetHash();
+    }
+    return ComputeMerkleRoot(std::move(leaves));
+}
+
+SignetTxs SignetTxs::Create(const CBlock& block, const CScript& challenge)
+{
+    CMutableTransaction tx_to_spend;
+    tx_to_spend.nVersion = 0;
+    tx_to_spend.nLockTime = 0;
+    tx_to_spend.vin.emplace_back(COutPoint(), CScript(OP_0), 0);
+    tx_to_spend.vout.emplace_back(0, challenge);
+
+    CMutableTransaction tx_spending;
+    tx_spending.nVersion = 0;
+    tx_spending.nLockTime = 0;
+    tx_spending.vin.emplace_back(COutPoint(), CScript(), 0);
+    tx_spending.vout.emplace_back(0, CScript(OP_RETURN));
+
+    // can't fill any other fields before extracting signet
+    // responses from block coinbase tx
+
+    // find and delete signet signature
+    if (block.vtx.empty()) return {tx_to_spend, tx_spending}; // no coinbase tx in block; invalid
+    CMutableTransaction cb(*block.vtx.at(0));
+
+    const int cidx = GetWitnessCommitmentIndex(block);
+    assert(cidx != NO_WITNESS_COMMITMENT);
+
+    CScript& script = cb.vout.at(cidx).scriptPubKey;
+
+    std::vector<uint8_t> data;
+    if (!ExtractCommitmentSection(script, SIGNET_HEADER, data)) {
+        return {tx_to_spend, tx_spending}; // no commitment; invalid
+    }
+    try {
+        VectorReader v(SER_NETWORK, INIT_PROTO_VERSION, data, 0);
+        v >> tx_spending.vin[0].scriptSig;
+        if (!v.empty()) v >> tx_spending.vin[0].scriptWitness.stack;
+        if (!v.empty()) return {tx_to_spend, tx_spending}; // extraneous data encountered; invalid
+    } catch (const std::exception&) {
+        return {tx_to_spend, tx_spending}; // excepted; invalid
+    }
+    uint256 signet_merkle = ComputeModifiedMerkleRoot(cb, block);
+
+    data.clear();
+    CVectorWriter writer(SER_NETWORK, INIT_PROTO_VERSION, data, 0);
+    writer << block.nVersion;
+    writer << block.hashPrevBlock;
+    writer << signet_merkle;
+    writer << block.nTime;
+    tx_to_spend.vin[0].scriptSig << data;
+    tx_spending.vin[0].prevout = COutPoint(tx_to_spend.GetHash(), 0);
+
+    return {tx_to_spend, tx_spending, true};
+}
+
+// Signet block solution checker
+bool CheckBlockSolution(const CBlock& block, const Consensus::Params& consensusParams)
+{
+    if (block.GetHash() == consensusParams.hashGenesisBlock) {
+        // genesis block solution is always valid
+        return true;
+    }
+
+    int cidx = GetWitnessCommitmentIndex(block);
+    if (cidx == NO_WITNESS_COMMITMENT) {
+        return error("CheckBlockSolution: Errors in block (no witness commitment)");
```

Comment:
E.g. in `ReadBlockFromDisk`, this isn't remote peer misbehavior, is it? Either way, it uses `error()` when errors in the block is encountered, and depends on `CheckBlockSolution` to do so as well. I could remove that dependency and just put a generic error in `ReadBlockFromDisk`, but not sure it's necessary.

The `error()` result is unused in `CheckBlock()`.

Reply:
What I meant to say is that `error()` will unconditionally log to disk (something that should happen on local system errors, e.g. ironically out-of-disk). Reject reasons that originate from messages sent by remote peers should be logged with a log category, which can be turned on or off.

Reply:
I'm confused. The other `error()` calls e.g. in `ReadBlockFromDisk` are exactly of the same type. If you mean that these are invalid and should (eventually) be fixed, then I'm completely on board.

The end result, in this case, will still be a log to disk, since `ReadBlockFromDisk` calls `error()` -- it will just be less detailed/helpful.

I could break the current pattern, i.e. not call `error()` on this particular failure type, but every other type does call `error()`, so it seems sensible. See 1152, 1160, 1165 below.

https://github.com/bitcoin/bitcoin/blob/a789f251e8fa2d1569c8f4d23f84d75bf74e1f94/src/validation.cpp#L1146-L1174

Reply:
error() is for system errors, situations where something is wrong with your computer locally, rather than something wrong with the block itself.

This here looks like a consensus rule violation instead.

Reply:
OK, I think L1164 is similar, but I can sort of get the difference. Changing to logging.

Reply:
@kallewoof Just to make sure this is clear: L1164 isn't part of the consensus rules, but just a disk-sanity check.

The reason is that we don't store blocks with invalid PoW on disk, as it's verified before that point is reached (it's verified in `CheckBlockHeader`). In `ReadBlockFromDisk` the check is redundant for that reason, but it's used as a simple (but rather expensive) checksum that the header is read correctly.

The code here (I believe?) is the actual signet consensus rule implementation, so it's more akin to what `CheckBlockHeader` is doing, and shouldn't unconditionally log violations.

Reply:
Good point, yep!

Reply:
I think this has been correctly solved in the latest force push (Thus I resolved the discussion here)